Sources/Core/EthereumABI/ABIElements.swift

```diff
@@ -264,72 +264,75 @@ extension ABI.Element.Function {
     }
 
-    public func decodeReturnData(_ data: Data) -> [String: Any]? {
-        // the response size greater than equal 100 bytes, when read function aborted by "require" statement.
-        // if "require" statement has no message argument, the response is empty (0 byte).
-        if data.bytes.count >= 100 {
-            let check00_31 = BigUInt("08C379A000000000000000000000000000000000000000000000000000000000", radix: 16)!
-            let check32_63 = BigUInt("0000002000000000000000000000000000000000000000000000000000000000", radix: 16)!
-
-            // check data[00-31] and data[32-63]
-            if check00_31 == BigUInt(data[0...31]) && check32_63 == BigUInt(data[32...63]) {
-                // data.bytes[64-67] contains the length of require message
-                let len = (Int(data.bytes[64])<<24) | (Int(data.bytes[65])<<16) | (Int(data.bytes[66])<<8) | Int(data.bytes[67])
-
-                let message = String(bytes: data.bytes[68..<(68+len)], encoding: .utf8)!
-
-                print("read function aborted by require statement: \(message)")
-
-                var returnArray = [String: Any]()
-
-                // set infomation
-                returnArray["_abortedByRequire"] = true
-                returnArray["_errorMessageFromRequire"] = message
-
-                // set empty values
-                for i in 0 ..< outputs.count {
-                    let name = "\(i)"
-                    returnArray[name] = outputs[i].type.emptyValue
-                    if outputs[i].name != "" {
-                        returnArray[outputs[i].name] = outputs[i].type.emptyValue
-                    }
-                }
+    /// Decodes data returned by a function call. Able to decode `revert("...")` calls.
+    /// - Parameter data: bytes returned by a function call.
+    /// - Returns: a dictionary containing returned data mappend to indices and names of returned values if these are not `nil`.
+    /// Return cases:
+    ///     - when no `outputs` declared: returning `["_success": true]`;
+    ///     - when `outputs` declared and decoding completed successfully: returning `["_success": true, "0": value_1, "1": value_2, ...]`.
+    ///     Additionally this dictionary will have mappings to output names if these names are specified in the ABI.
+    ///     - function call was aborted using `require(some_string_error_message)`: returning `["_success": false, "_abortedByRequire": true, "_errorMessageFromRequire": error_message]`.
+    ///     - in case of any error: returning `["_success": false, "_failureReason": String]`;
+    ///         Error reasons include:
+    ///            -  `outputs` declared but at least one value failed to be decoded;
+    ///            - `data.count` is less than `outputs.count * 32`;
+    ///            - `outputs` defined and `data` is empty;
+    ///            - `data` represent reverted transaction
+    public func decodeReturnData(_ data: Data) -> [String: Any] {
```

Comment:
What should we also consider here is the option to pass a list or a dictionary of `ABI.Element.EthError`.
The reason for that is the ability to use custom errors with `revert`, e.g. `revert Unauthorized()`.
That list must be optional so that we don't force anyone to set it in order to decode data but as a result of having that list we can decode custom errors that were raised using `revert` because the first 4 bytes of the given data will match function selector of a custom error.

Reply:
`revert` test cases:
```
revert() -> 0 bytes. We can only assume it was a result of `revert()` call. 

revert Unauthorized() -> 0x82b4290000000000000000000000000000000000000000000000000000000000

revert("Not enough Ether provided.") ->
08c379a0
0000000000000000000000000000000000000000000000000000000000000020
000000000000000000000000000000000000000000000000000000000000001a
4e6f7420656e6f7567682045746865722070726f76696465642e000000000000

revert Unauthorized("Reason") -> 
973d02cb
0000000000000000000000000000000000000000000000000000000000000020
0000000000000000000000000000000000000000000000000000000000000006
526561736f6e0000000000000000000000000000000000000000000000000000
```

Reply:
`require` test cases:
```
require(false) -> 0 bytes. We can only assume it was a result of `revert()` or `require(expression)` call.

require(false, "Not enough Ether provided.") ->
08c379a0
0000000000000000000000000000000000000000000000000000000000000020
000000000000000000000000000000000000000000000000000000000000001a
4e6f7420656e6f7567682045746865722070726f76696465642e000000000000
```

Reply:
`require` is unable to use custom errors so there's no third test case:
> It is currently not possible to use custom errors in combination with `require`. Please use `if (!condition) revert CustomError();` instead.
https://docs.soliditylang.org/en/v0.8.17/control-structures.html?highlight=require()#panic-via-assert-and-error-via-require